### ISSUE TITLE
Add localization support to GitLens

### ISSUE THREAD

**Issue by eamodio** [eamodio/vscode-gitlens#577]
Here is a [sample extension](https://github.com/Microsoft/vscode-extension-samples/tree/master/i18n-sample) using the [vscode-nls](https://github.com/Microsoft/vscode-nls) library

**Comment by axetroy** [comment]
I have tried to implement i18n with `vscode-nls`.

Then encountered a lot of difficulties

I found that `vscode-nls` is not so friendly.

here we go:

**1. `vscode-nls` is not a standalone package**

We have to use `Transifex` as a third-party hosting service that hosts i18n files.

If we need to modify the translation file, we need to log in `Transifex` and add a collaborator to modify it.

This undoubtedly prevents developers from further contributing to translation.

In order to integrate `Transifex`

We have to use `Gulp`, which is not very compatible with the workflow

**2. `official i18n example` documenation not clear**

How to write i18n file?

**3. `vscode-nls` API too complicated**

What did this function do?

That's all!

I I disapprove follow `offcial i18n example` to implement i18n.

**Comment by eamodio** [comment]
While I agree that the official i18n sample is seemingly over complicated and the documentation is not very clear, the `vscode-nls` library which is standalone as far as can see, seems straight forward. I don't see where `vscode-nls` has any dependency on `Transifex` -- where is that?

Also will the files that have the localized strings be in the same format in what you are proposing vs `vscode-nls`?

**Comment by eamodio** [comment]
The examples you've posted are full extensions -- what are the pieces (and dependencies if any) that would be brought into GitLens?

**Comment by axetroy** [comment]
`vscode-nls` does not matter with `Transifex`.

But the official i18n example uses it. If you use `vscode-nls` alone, you have to explore it yourself.

The examples I've posted is the solution. not a library.

localized strings format same with official solution

You can see here:

English by default:

https://github.com/axetroy/vscode-gpm/blob/a00147fc2bc0e85296f65dcdf7c09a2dfba60cf1/package.nls.json#L1-L10

Chinese Simplified

https://github.com/axetroy/vscode-gpm/blob/a00147fc2bc0e85296f65dcdf7c09a2dfba60cf1/package.nls.zh-cn.json#L1-L10

**Comment by eamodio** [comment]
OK, sounds good -- thanks for the information. FYI, will also need to make sure the solution works well with webpack -- which will likely mean avoiding the use of `require` to load json files on disk -- instead use `readFileSync`

**Comment by axetroy** [comment]
@eamodio

Yes, the project must have the translation files.

Add copyPlugin to webpack to copy the translation file to the dist directory.

Because there is too much place to translate.

I suggest that the first stage of the translation of the command / configuration just fine.

**Comment by eamodio** [comment]
@axetroy works for me! Thanks again! Looking forward to the PR

**Comment by axetroy** [comment]
@eamodio 
Yes. I am working on it.

This will take a lot of time due to a large number of configuration items.

**Comment by chouzz** [comment]
Actually，there is a more convenient way to implement i18n.  Microsoft already has a tool called [vscode-nls-dev](https://github.com/microsoft/vscode-nls-dev) , this tool automates the extraction of strings to be externalized from TS and JS code.  And you can also add localization using [Transifex](www.transifex.com).

**Comment by eamodio** [comment]
@chouzz PR is very welcome 😄

**Comment by goldst** [comment]
@chouzz are you working on this? If not, I'd take a look.

**Comment by ImRodry** [comment]
If you do add localization, please do so through [Crowdin](https://crowdin.com)

**Comment by goldst** [comment]
Quick update: I almost finshed a basic i18n system that uses vscode-nls-dev / webpack, but unfortunately I was quite occupied with other things in the last weeks. I hope to finish it next week.

vscode-nls-dev does not necessarily require a localization platform like Transifex (or Crowdin). In the most basic version one could just edit the json files that contain translations, and that is how my PR will work. I suggest that selecting and implementing the usage of a platform should follow in a separate PR.

**Comment by ImRodry** [comment]
Editing json files directly through a PR is definitely not an ideal option. If you want many users helping out you will need to set up a Crowdin project otherwise the idea will most likely die pretty fast

**Comment by goldst** [comment]
@ImRodry Yes, I am not saying that this project shouldn't use Crowdin or something comparable. I wanted to say that it is the second step and therefore should only be done after the first step - implementing vscode-nls-dev - is finished.

**Comment by ImRodry** [comment]
But I think it should all be put in the same PR since the only thing crowdin needs is a crowdin.yml file on the repo and the rest is handled by the managers of the project on Crowdin itself. The configuration file should also be pretty simple

**Comment by pgy866** [comment]
非常希望能得到本地化的支持，这个插件非常棒

**Comment by airaketa** [comment]
Just FYI, I've made a [pull request](https://github.com/gitkraken/vscode-gitlens/pull/1937) introducing i18n support. I expect it may take some time until it's merged.

**Comment by zWingz** [comment]
https://github.com/microsoft/vscode-loc

**Comment by xujz520** [comment]
Excuse me, how is the progress, looking forward to supporting multiple languages as soon as possible

**Comment by Xavier9896** [comment]
I'm also looking forward to support for Chinese!

**Comment by uuiid** [comment]
I'm also looking forward to support for Chinese!

**Comment by erickisos** [comment]
Hey @eamodio!

I see that there is still interest to getting this done, is there anything we can do to agilize this process? I'm looking forward to help with some of the translations, but would love to understand if from your perspective this is something that can't be achieved in the upcoming months.

**Comment by uuiid** [comment]
If possible, I can also provide some assistance

**Comment by AkazaAkaria** [comment]
我也期待着对中文的支持！

**Comment by acmuhan** [comment]
When is support for Chinese expected?

**Comment by Xuaner183** [comment]
We need Chinese

**Comment by Y-PLONI** [comment]
?

**Comment by xuhuanzy** [comment]
纯 vibecode 出来的, 使用编译期分析Ast再生成方案, 对`./src`与`package.json`的侵入性为0, 可以持续合并上游. 
且将翻译部分抽取为单独的模块, 虽然目前99%翻译是`codex`做的, 但可以很方便修正翻译.

https://github.com/xuhuanzy/vscode-gitlens-zh

另外我不觉得官方会提供本地化版本, 他里面有巨量 html 页面内容是零碎的, 还有巨多神秘动态拼接字符串, 最多给你翻译下右键菜单意思下得了

**Comment by eamodio** [comment]
Update: GitLens now has full localization support. It's in tomorrow's pre-release and will be in the next stable release.

What's in it:
- Every user-facing string (commands, settings, notifications, quick picks, tree views, hovers, and all webviews including the Commit Graph) goes through VS Code's localization API and follows the display language you've set in VS Code. There's no separate GitLens language setting.
- Plural forms use CLDR categories, so languages with more than two forms (Russian, Polish, Arabic, …) render correctly.
- Three experimental languages ship to start: **Spanish**, **Simplified Chinese** and **Traditional Chinese**.

One caveat: these three translations were AI-generated and AI-reviewed. No native speaker has been through them yet. Treat them as a strong first draft. Wording, terminology and tone will have rough edges, and that's exactly where we'd love the community's help. If you're on the pre-release, switch VS Code's display language to one of the three and tell us what reads wrong.

How to help:
- **Improve an existing language**: the catalogs are plain JSON, one entry per message, in `l10n/bundle.l10n.<locale>.json` and `package.nls.<locale>.json`. Fix the entries you disagree with and open a PR. Small PRs are welcome; you don't need to review the whole file.
- **Add a new language**: `l10n/README.md` in the repo explains the workflow, including tooling that splits the untranslated messages into small chunks, the terminology conventions, and how plural messages are written. Start from that guide.
- **Review**: if you're a native speaker of one of the three shipped languages and want to do a proper pass, say so here and we'll coordinate so people don't overlap.

One rule for all of it: every new or changed language needs human review before we advertise it as supported. AI drafts are fine as a starting point (that's how the first three were made), but a native speaker has to sign off.

Thanks for the patience on this one, and to everyone who offered to help over the years. Now there's something concrete to help with.
